File: rust/src/engine.rs
use std::fmt::Debug;
use std::future::Future;
use std::pin::Pin;
use futures_util::future::{Abortable, AbortHandle};
use log::{debug, info};
use tokio::time::sleep;
use tower::Service;
use crate::data::httpapi::alerts::{Alert, AlertMessage, AlertName, AlertState, QueryName};
use crate::data::httpapi::{AlertStatusChangeRequest, NotifyRequest, QueryAlertsResponse, QueryResponse, ResolveRequest};

const LOG_TARGET: &str = "alerts_engine";
fn log_target_alert(alert: &AlertName) -> String {
    format!("{}::{}", LOG_TARGET, alert.0)
}

pub async fn initialize<
    Err : 'static + Send + Debug,
    DoQuery : 'static + Service<QueryName, Error = Err, Response = QueryResponse> + Clone + Send,
    Notify : 'static + Service<NotifyRequest, Error = Err, Response = ()> + Clone + Send,
    Resolve : 'static + Service<ResolveRequest, Error = Err, Response = ()> + Clone + Send,
>(
    mut query_alerts: impl Service<(), Error = Err, Response = QueryAlertsResponse>,
    query: DoQuery, notify: Notify, resolve: Resolve
) -> Result<(), Err>
    where
        <DoQuery as Service<QueryName>>::Future : Send,
        <Notify as Service<NotifyRequest>>::Future : Send,
        <Resolve as Service<ResolveRequest>>::Future : Send,
{
    info!(target: LOG_TARGET, "Querying alerts...");
    let alerts: QueryAlertsResponse = query_alerts.call(()).await?;
    info!(target: LOG_TARGET, "Alerts queried:\n{:?}", alerts);

    let launched_alerts =
        alerts.alerts.into_iter()
            .map(|alert| (
                alert.name.clone(),
                tokio::spawn(initialize_alert(
                    alert, query.clone(), notify.clone(), resolve.clone()
                ))
            ))
            .collect::<Vec<_>>();

    for (alert_name, join_handle) in launched_alerts {
        let res = join_handle.await;
        info!(target: log_target_alert(&alert_name).as_str(), "Alert engine finished with {:?}", res);
    }

    Ok(())
}

/// Watches a single [Alert].
pub async fn initialize_alert<Err : Send + 'static, DoNotify, DoResolve>(
    alert: Alert,
    mut query: impl Service<QueryName, Error = Err, Response = QueryResponse>,
    notify: DoNotify,
    resolve: DoResolve,
) -> Result<(), Err>
where
    DoNotify : Service<NotifyRequest, Error = Err, Response = ()> + Clone + Send + 'static,
    <DoNotify as Service<NotifyRequest>>::Future : Send,
    DoResolve : Service<ResolveRequest, Error = Err, Response = ()> + Clone + Send + 'static,
    <DoResolve as Service<ResolveRequest>>::Future : Send
{
    let log_target = log_target_alert(&alert.name);
    let create_notify_request = |message: AlertMessage| {
        AlertStatusChangeRequest::Notify {
            request: NotifyRequest { alert_name: alert.name.clone(), message },
            repeat_after: alert.repeat_interval.clone()
        }
    };

    info!(target: &log_target, "Starting engine.");
    let mut state = AlertState::Pass;
    let mut maybe_reporter: Option<AbortHandle> = None;
    loop {
        debug!(target: &log_target, "Querying {:?}", alert.name);
        let response: QueryResponse = query.call(alert.query.clone()).await?;
        debug!(target: &log_target, "Received response: {:?}", response.value);
        let new_state = alert.thresholds.state_for(&response.value.into());
        if state == new_state {
            debug!(target: &log_target, "State did not change");
        }
        else {
            info!(target: &log_target, "State changed from {:?} to {:?}", state, new_state);
            state = new_state;

            if let Some(reporter) = &maybe_reporter {
                debug!(target: &log_target, "Detected previous reporter, stopping it.");
                reporter.abort();
            }
            maybe_reporter = Some(initialize_alert_reporter(
               alert.name.clone(),
                match state {
                    AlertState::Pass => AlertStatusChangeRequest::Resolve(ResolveRequest {
                        alert_name: alert.name.clone()
                    }),
                    AlertState::Warn { message } =>
                        create_notify_request(message.clone()),
                    AlertState::Critical { message } =>
                        create_notify_request(message.clone())
                },
                notify.clone(), resolve.clone()
            ));
        }
        debug!(target: &log_target, "Sleeping for {:?}", alert.interval);
        sleep(alert.interval).await;
    }
}

/// Initializes the reporter task, which runs in the background and can be stopped using the
/// returned [AbortHandle].
pub fn initialize_alert_reporter<Err : Send + 'static, DoNotify, DoResolve>(
    alert_name: AlertName,
    request: AlertStatusChangeRequest,
    mut notify: DoNotify,
    mut resolve: DoResolve,
) -> AbortHandle
    where
        DoNotify : Service<NotifyRequest, Error = Err, Response = ()> + Send + 'static,
        <DoNotify as Service<NotifyRequest>>::Future : Send,
        DoResolve : Service<ResolveRequest, Error = Err, Response = ()> + Send + 'static,
        <DoResolve as Service<ResolveRequest>>::Future : Send
{
    let mut log_target = log_target_alert(&alert_name);
    log_target.push_str("::reporter");
    let task: Pin<Box<dyn Future<Output = Result<(), Err>> + Send>> = match request {
        AlertStatusChangeRequest::Resolve(resolve_request) => {
            // Invoke the resolve once.
            let future = resolve.call(resolve_request);
            // Pin the future for the `spawn`.
            Box::pin(future)
        }
        AlertStatusChangeRequest::Notify { request, repeat_after } => {
            // Invoke the notify, then repeat every interval.
            let future = async move {
                loop {
                    debug!(target: &log_target, "Sending notify: {:?}", request);
                    notify.call(request.clone()).await?;
                    debug!(target: &log_target, "Sleeping next notify for {:?}", repeat_after);
                    sleep(repeat_after).await;
                }
            };
            // Pin the future for the `spawn`.
            Box::pin(future)
        }
    };
    let (abort_handle, abort_registration) = AbortHandle::new_pair();
    tokio::spawn( Abortable::new(task, abort_registration));
    // Return the abort handle so that we could stop the spawned task.
    abort_handle
}

File: go/engine/engine.go
package engine

import (
	"alert-execution-engine/data/httpapi"
	. "alert-execution-engine/functional"
	"alert-execution-engine/utils"
	"go.uber.org/zap"
	"sync"
	"time"
)

// DoNotify takes in isAborted because we need to communicate to the HTTP api that we should stop retrying sending a
// request. If it fails, the reporter stops.
type DoNotify func(isAborted utils.IsChannelClosed, request httpapi.NotifyRequest) error

// DoResolve - same as DoNotify.
type DoResolve func(isAborted utils.IsChannelClosed, request httpapi.ResolveRequest) error

func Initialize[Err error](
	log zap.Logger,
	// This function could be a generic type, but Goland freaks out :)
	queryAlerts func() (*httpapi.QueryAlertsResponse, *Err),
	// DoQuery can fail, if it does, the alert engine for a particular alert will die.
	doQuery func(name httpapi.QueryName) (*httpapi.QueryResponse, *Err),
	doNotify DoNotify,
	doResolve DoResolve,
) *Err {
	log.Info("Querying alerts...")
	alerts, err := queryAlerts()
	if err != nil {
		return err
	}
	log.Info("Alerts queried.", zap.Any("alerts", alerts))

	waitGroup := new(sync.WaitGroup)
	waitGroup.Add(len(alerts.Alerts))
	for _, alert := range alerts.Alerts {
		go InitializeAlerts(log, waitGroup, alert, doQuery, doNotify, doResolve)
	}

	log.Info("Waiting until all alert engines finish or error out.")
	waitGroup.Wait()

	return nil
}

func InitializeAlerts[Err error](
	log zap.Logger,
	waitGroup *sync.WaitGroup,
	alert httpapi.Alert,
	// DoQuery can fail, if it does, the alert engine for a particular alert will die.
	doQuery func(name httpapi.QueryName) (*httpapi.QueryResponse, *Err),
	doNotify DoNotify,
	doResolve DoResolve,
) *Err {
	// If we error out, notify the master goroutine.
	defer waitGroup.Done()

	alertLogField := zap.String("alert", alert.Name.String())
	log.Info("Starting engine.", alertLogField)

	var state httpapi.AlertState = &httpapi.AlertStatePass{}
	var maybeReporter = None[RunningReporter]()

	for {
		log.Debug("Querying", alertLogField)
		response, err := doQuery(alert.Query)
		if err != nil {
			log.Error("Querying failed", alertLogField, zap.Error(*err))
			return err
		}
		log.Debug("Received response", alertLogField, zap.Any("response", response))
		newState := alert.Thresholds.StateFor(response.Value.ToAlertThresholdValue())
		if state == newState {
			log.Debug("State did not change", alertLogField)
		} else {
			log.Info("State changed", alertLogField, zap.Any("from", state), zap.Any("to", newState))
			state = newState

			maybeReporter.IfSome(func(reporter RunningReporter) {
				log.Debug("Detected previous reporter, stopping it.", alertLogField)
				reporter.abort()
			})
			statusChangeRequest := Fold[httpapi.AlertMessage, httpapi.AlertStatusChangeRequest](
				state.NotifyMessage(),
				func() httpapi.AlertStatusChangeRequest { return &httpapi.ResolveRequest{AlertName: alert.Name} },
				func(message httpapi.AlertMessage) httpapi.AlertStatusChangeRequest {
					return &httpapi.AlertStatusChangeNotifyRequest{
						Request: httpapi.NotifyRequest{
							AlertName: alert.Name,
							Message:   message,
						},
						RepeatAfter: alert.RepeatInterval,
					}
				},
			)
			maybeReporter = Some(InitializeAlertReporter(log, alert.Name, statusChangeRequest, doNotify, doResolve))

			log.Debug("Sleeping...", alertLogField, zap.Duration("for", alert.Interval))
			time.Sleep(alert.Interval)
		}
	}
}

type RunningReporter struct {
	// Abort the reporter goroutine by closing this channel
	abortChannel chan Unit
}

func (self *RunningReporter) abort() {
	close(self.abortChannel)
}

func InitializeAlertReporter[Err any](
	log zap.Logger,
	name httpapi.AlertName,
	request httpapi.AlertStatusChangeRequest,
	doNotify DoNotify,
	doResolve DoResolve,
) RunningReporter {
	abortChannel := make(chan Unit)
	isAborted := utils.ProduceCheckIfChanIsClosedFunc(abortChannel)
	go alertReporterMain(
		log, name, request,
		func(request httpapi.NotifyRequest) error { return doNotify(isAborted, request) },
		func(request httpapi.ResolveRequest) error { return doResolve(isAborted, request) },
	)

	return RunningReporter{abortChannel: abortChannel}
}

func alertReporterMain(
	log zap.Logger,
	name httpapi.AlertName,
	request httpapi.AlertStatusChangeRequest,
	doNotify func(request httpapi.NotifyRequest) error,
	doResolve func(request httpapi.ResolveRequest) error,
) {
	alertLogField := zap.String("alert-reporter", name.String())

	request.ToEither().Match(
		func(req httpapi.AlertStatusChangeNotifyRequest) {
			// Invoke doNotify, then repeat every interval.
			for {
				log.Debug("Sending notify", alertLogField, zap.Any("request", req.Request))
				err := doNotify(req.Request)
				if err != nil {
					log.Error("Notify failed!", alertLogField, zap.Error(err))
					return
				}
				log.Debug(
					"Sleeping until next notify", alertLogField, zap.Duration("sleepFor", req.RepeatAfter),
				)
				time.Sleep(req.RepeatAfter)
			}
		},
		func(req httpapi.ResolveRequest) {
			// Invoke the resolve once.
			err := doResolve(req)
			if err != nil {
				log.Error("Resolve failed!", alertLogField, zap.Error(err))
			}
		},
	)
}
